### ISSUE TITLE
Upgrade on Windows results in loss of taskbar icon

### ISSUE THREAD

**Issue by xpusostomos** [keepassxreboot/keepassxc#7576]
## Overview

When you upgrade KeepassXC on Windows, you lose the pinned icon from the windows taskbar.

This has been reported before, and alleged to be fixed, but the KeepassXC just released (2.7.0) still has this issue

https://github.com/keepassxreboot/keepassxc/issues/4226

I don't know if it's a Windows 11 thing, but this seems even more annoying than it used to be because if your whole taskbar is full, Windows won't let you pin new things to it (even though some of those open apps might not be pinned). The only way I've found to easily get around that is to scale the display back to 100% (too bad if it was already 100% I guess), then I can pin the app again to the taskbar, move it back to where it should be, then rescale back to 125% or whatever.

OK, sure windows is garbage, but other apps don't seem to lose their taskbar pinned icons on upgrade, and it sure would be nice if KeepassXC didn't either.

Operating System: Windows

**Comment by droidmonkey** [comment]
I've spent literal hours trying to fix this, nothing worked. I am not very keen on spending more hours on this.

**Comment by xpusostomos** [comment]
Fair enough I guess. But it would be good if you could document what sort of things you tried... that might help someone else who has ideas... or someone might suggest something.

**Comment by mooms06** [comment]
It's a very minor issue IMO, and the fix is quick and easy: just unpin the old broken icon and pin the new one.

**Comment by phoerious** [comment]
This affects many other applications as well. It might be a general Windows bug.

**Comment by mooms06** [comment]
In my experience it don't affect many applications, on all my pinned applications, IIRC I've only encountered it with a portable version of WhatsApp and KeepassXC. I'm on Windows 10.

**Comment by droidmonkey** [comment]
It is a common problem with applications installed through an msi. All provided solutions on the interwebs fail to fix the problem. So, either we are doing something very wrong with our shortcut icons... or there is a bug somewhere in the install process itself.

**Comment by mooms06** [comment]
I've installed 2.6.6 , pinned the icon to the taskbar, then updated to 2.7.0, and here is my findings:

After updating, the shortcut was moved from
`%APPDATA%\Microsoft\Internet Explorer\Quick Launch\User Pinned\TaskBar `
to 
`%APPDATA%\Microsoft\Internet Explorer\Quick Launch\User Pinned\TaskBar\Tombstones`

![image](https://user-images.githubusercontent.com/25405517/159508146-4519a087-d712-42b3-9853-2ccd3deaebef.png)

If we move it in the original folder, this is not enough to make the icon reappear, we have to rebuild (reset) the icon cache, (I used [Rebuild Shell Icon Cache ](https://www.sordum.org/9194/rebuild-shell-icon-cache-v1-3/)), then the icon is back:

![image](https://user-images.githubusercontent.com/25405517/159509303-c4861978-4901-4a69-932a-39acca1146c1.png)



So, the fix could be either:

- Do not move the shortcut (the best solution IMO).

- If it's not possible, move back the shortcut to the correct location, then rebuild the icon cache (can be done without admin rights).


Edit: the pinned icons are also referenced in
`HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer\Taskband`

There is no need to touch it if the icon was pinned before the update, but the entries need to be populated for the pinned icons to show, placing a lnk file (a shortcut) in the folder `%APPDATA%\Microsoft\Internet Explorer\Quick Launch\User Pinned\TaskBar\`  is not enough.

**Comment by phoerious** [comment]
> In my experience it don't affect many applications, on all my pinned applications, IIRC I've only encountered it with a portable version of WhatsApp and KeepassXC. I'm on Windows 10.

I regularly have this with all JetBrains and Adobe products.

**Comment by mooms06** [comment]
Then you could probably fix it by moving the shortcut(s) as detailed above.

**Comment by xpusostomos** [comment]
I'm very ignorant about Windows installation, other than to observe that Microsoft sucks, but....

* Should we, could we... make keepassxc available in the Microsoft store, and would this help?
* Microsoft was making fanfare recently about their new package management system (the windows equivilent of Linux apt or yum or dpkg). In theory you can install or update any app from the command line as easily as on Linux. Could we do that, should we do that? https://docs.microsoft.com/en-us/windows/package-manager/

**Comment by mooms06** [comment]
That's unrelated to the issue here.

**Comment by phoerious** [comment]
MS store presence will probably not fix this. Whether we can do it is more like a decision of whether we can stand being the second and only other non-spam app in the store after "Microsoft Terminal".

**Comment by Jon-guy30** [comment]
Is there any particular reason why this software is distributed as an MSI package only? Most software is distributed as an EXE and in some cases .exe and .msi depending on what one wants to choose. I'm just curious.

**Comment by droidmonkey** [comment]
.msi is the defacto standard for installers on windows.

**Comment by Jon-guy30** [comment]
Wow, never knew that. Thanks!

**Comment by mooms06** [comment]
It's not on the list of the fixed issues, but after upgrading to 2.7.1.0, I didn't get the blank taskbar icon this time, has this been fixed or is it a coincidence ?

Edit: made some tests: 

- Upgrading to 2.7.1 from 2.7.0 icon still here, as stated.

- Upgrading to 2.7.1 from 2.6.6, blank icon, so apparently the fix was baked in 2.7.0, but only for future updates ?

**Comment by droidmonkey** [comment]
Hey that might be true 👍

**Comment by droidmonkey** [comment]
Haven't run into this for a year
